src/uopool/services/uopool.rs

```diff
@@ -30,12 +33,38 @@ impl UoPoolService {
 impl UoPool for UoPoolService {
     async fn add(
         &self,
-        _request: tonic::Request<AddRequest>,
+        request: tonic::Request<AddRequest>,
     ) -> Result<Response<AddResponse>, tonic::Status> {
-        // let req = request.into_inner();
-        // TODO: sanity checks
-        // TODO: simulation
-        Err(tonic::Status::unimplemented("todo"))
+        let req = request.into_inner();
+        let mut res = AddResponse::default();
+
+        if let Some(user_operation) = req.uo {
+            let user_operation: UserOperation = user_operation
+                .try_into()
+                .map_err(|_| tonic::Status::invalid_argument("invalid user operation"))?;
+
+            info!("{:?}", user_operation);
+
+            // TODO: validate user operation
+            // TODO: sanity checks
+            // TODO: simulation
+
+            let uo_pool_error = UoPoolError::owned(
+                -32602,
+                "user operation was not added",
+                Some(json!({
+                    "reason": "this is error",
+                })),
+            );
+
+            res.set_result(AddResult::NotAdded);
+            res.data = serde_json::to_string(&uo_pool_error)
+                .map_err(|_| tonic::Status::internal("error adding user operation"))?;
+
+            return Ok(tonic::Response::new(res));
```

Comment:
Although I got questions with the codes above, I think because it is not complete yet.

Reply:
Yes, that's right, this part is not yet completely done.